File: eveil/reference/wordend/dsp.py
"""Primitives DSP en Python stdlib pur — la référence de `WordEndCore` (Swift).

Règle de portabilité : chaque fonction a son équivalent Swift **ligne à ligne**
(`eveil/ios/Sources/WordEndCore/`). Pas de numpy, aucun aléa non seedé, des
constantes explicites. Les deux implémentations sont tenues en parité par des
vecteurs de référence (`golden.py` → `golden_vectors.json`).

Le signal est analysé à une fréquence **canonique de 24 kHz** : assez pour les
sibilantes d'enfant (énergie jusque vers 10 kHz), deux fois moins coûteux que
les 48 kHz natifs de l'iPad (qui sont décimés par 2 avant analyse).
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Sequence

ANALYSIS_RATE = 24_000          # Hz — fréquence d'analyse canonique
FRAME_SIZE = 512                # échantillons = taille FFT (21,3 ms à 24 kHz)
HOP_SIZE = 240                  # échantillons (10 ms à 24 kHz)
BIN_HZ = ANALYSIS_RATE / FRAME_SIZE   # 46,875 Hz — exactement représentable
EPS = 1e-12                     # plancher numérique avant log10
MASK64 = 0xFFFF_FFFF_FFFF_FFFF


def frame_ms() -> float:
    """Pas d'analyse en millisecondes (10 ms)."""
    return 1000.0 * HOP_SIZE / ANALYSIS_RATE


def ms_to_frames(ms: float) -> int:
    """Convertit une durée en nombre de trames (arrondi au plus proche)."""
    return int(round(ms / frame_ms()))


# ---------------------------------------------------------------------------
# Générateur pseudo-aléatoire déterministe (identique bit à bit en Swift)
# ---------------------------------------------------------------------------

class SplitMix64:
    """SplitMix64 (Steele, Lea & Flood 2014) — trivial à porter à l'identique.

    En Swift, les additions/multiplications utilisent `&+` / `&*` (modulo 2⁶⁴) ;
    ici, on masque explicitement sur 64 bits.
    """

    def __init__(self, seed: int) -> None:
        self.state = seed & MASK64

    def next_u64(self) -> int:
        self.state = (self.state + 0x9E3779B97F4A7C15) & MASK64
        z = self.state
        z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK64
        z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK64
        return z ^ (z >> 31)

    def next_unit(self) -> float:
        """Uniforme dans [0, 1) sur 53 bits de mantisse."""
        return (self.next_u64() >> 11) * (1.0 / 9007199254740992.0)

    def next_signed(self) -> float:
        """Uniforme dans [-1, 1) — bruit blanc sans fonction transcendante."""
        return 2.0 * self.next_unit() - 1.0


# ---------------------------------------------------------------------------
# FFT radix-2 itérative (portable ; le backend Accelerate est optionnel en Swift)
# ---------------------------------------------------------------------------

_TWIDDLES: dict[int, list[complex]] = {}


def _twiddles(n: int) -> list[complex]:
    table = _TWIDDLES.get(n)
    if table is None:
        table = [complex(math.cos(-2.0 * math.pi * k / n), math.sin(-2.0 * math.pi * k / n))
                 for k in range(n // 2)]
        _TWIDDLES[n] = table
    return table


def fft(values: Sequence[float]) -> list[complex]:
    """DFT d'un signal réel par Cooley-Tukey radix-2 (taille puissance de 2)."""
    n = len(values)
    if n == 0 or n & (n - 1):
        raise ValueError(f"taille FFT non puissance de 2 : {n}")
    a = [complex(v, 0.0) for v in values]
    j = 0
    for i in range(1, n):                     # permutation bit-reverse
        bit = n >> 1
        while j & bit:
            j ^= bit
            bit >>= 1
        j ^= bit
        if i < j:
            a[i], a[j] = a[j], a[i]
    table = _twiddles(n)
    size = 2
    while size <= n:
        half = size // 2
        step = n // size
        for start in range(0, n, size):
            for k in range(half):
                w = table[k * step]
                u = a[start + k]
                v = a[start + k + half] * w
                a[start + k] = u + v
                a[start + k + half] = u - v
        size *= 2
    return a


def dft_naive(values: Sequence[float]) -> list[complex]:
    """DFT directe O(n²) — sert uniquement à vérifier `fft` dans les tests."""
    n = len(values)
    out = []
    for k in range(n):
        acc = 0j
        for t, v in enumerate(values):
            ang = -2.0 * math.pi * k * t / n
            acc += v * complex(math.cos(ang), math.sin(ang))
        out.append(acc)
    return out


HANN = [0.5 - 0.5 * math.cos(2.0 * math.pi * i / FRAME_SIZE) for i in range(FRAME_SIZE)]


def band_bins(lo_hz: float, hi_hz: float) -> tuple[int, int]:
    """Indices de bins [k0, k1) couvrant la bande [lo_hz, hi_hz)."""
    k0 = max(1, math.ceil(lo_hz / BIN_HZ))
    k1 = min(FRAME_SIZE // 2 + 1, math.ceil(hi_hz / BIN_HZ))
    return k0, max(k0, k1)


def to_db(power: float) -> float:
    return 10.0 * math.log10(power + EPS)


# ---------------------------------------------------------------------------
# Descripteurs par trame
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class Bands:
    """Bandes d'analyse (Hz). Valeurs par défaut justifiées dans docs/EVEIL.md §5."""

    low: tuple[float, float] = (300.0, 2500.0)       # voyelles (F1/F2)
    high: tuple[float, float] = (3000.0, 11000.0)    # bruit de friction sibilant
    centroid: tuple[float, float] = (1000.0, 11000.0)


@dataclass(frozen=True)
class FrameFeatures:
    index: int
    rms_db: float        # énergie totale (dB, référence arbitraire)
    low_db: float        # énergie bande « voyelle »
    high_db: float       # énergie bande « friction »
    high_ratio: float    # E_high / (E_low + E_high) ∈ [0, 1]
    centroid_hz: float   # centroïde spectral sur la bande `centroid`
    zcr_hz: float        # passages par zéro par seconde
    peak: float          # max |x| (détection d'écrêtage)


def frame_count(n_samples: int) -> int:
    """Nombre de trames entièrement contenues dans le signal."""
    if n_samples < FRAME_SIZE:
        return 0
    return 1 + (n_samples - FRAME_SIZE) // HOP_SIZE


def compute_frame(samples: Sequence[float], index: int, bands: Bands = Bands()) -> FrameFeatures:
    """Descripteurs de la trame `index` (fenêtre de Hann, DC retiré)."""
    start = index * HOP_SIZE
    seg = list(samples[start:start + FRAME_SIZE])
    if len(seg) < FRAME_SIZE:
        seg.extend([0.0] * (FRAME_SIZE - len(seg)))
    mean = sum(seg) / FRAME_SIZE
    x = [s - mean for s in seg]
    energy = sum(v * v for v in x) / FRAME_SIZE

    crossings = 0
    prev_neg = x[0] < 0.0
    for v in x[1:]:
        neg = v < 0.0
        if neg != prev_neg:
            crossings += 1
        prev_neg = neg
    zcr_hz = crossings * ANALYSIS_RATE / FRAME_SIZE
    peak = max(abs(s) for s in seg)

    spectrum = fft([x[i] * HANN[i] for i in range(FRAME_SIZE)])
    power = [c.real * c.real + c.imag * c.imag for c in spectrum[:FRAME_SIZE // 2 + 1]]

    l0, l1 = band_bins(*bands.low)
    h0, h1 = band_bins(*bands.high)
    c0, c1 = band_bins(*bands.centroid)
    e_low = sum(power[l0:l1])
    e_high = sum(power[h0:h1])
    e_c = 0.0
    e_cf = 0.0
    for k in range(c0, c1):
        e_c += power[k]
        e_cf += power[k] * k * BIN_HZ
    centroid = e_cf / e_c if e_c > EPS else 0.0
    ratio = e_high / (e_low + e_high) if (e_low + e_high) > EPS else 0.0
    return FrameFeatures(index, to_db(energy), to_db(e_low), to_db(e_high), ratio,
                         centroid, zcr_hz, peak)


def compute_frames(samples: Sequence[float], bands: Bands = Bands()) -> list[FrameFeatures]:
    return [compute_frame(samples, t, bands) for t in range(frame_count(len(samples)))]


# ---------------------------------------------------------------------------
# Outils statistiques et signaux
# ---------------------------------------------------------------------------

def percentile(values: Sequence[float], p: float) -> float:
    """Percentile « rang inférieur » déterministe (p ∈ [0, 1])."""
    if not values:
        raise ValueError("percentile d'une liste vide")
    ordered = sorted(values)
    idx = int(p * (len(ordered) - 1))
    return ordered[max(0, min(len(ordered) - 1, idx))]


def moving_average(values: Sequence[float], width: int) -> list[float]:
    """Moyenne glissante centrée ; aux bords, moyenne des valeurs disponibles."""
    half = width // 2
    n = len(values)
    out = []
    for t in range(n):
        lo = max(0, t - half)
        hi = min(n, t + half + 1)
        out.append(sum(values[lo:hi]) / (hi - lo))
    return out


def estimate_f0(samples: Sequence[float], frame_index: int,
                f0_min: float = 90.0, f0_max: float = 800.0,
                min_corr: float = 0.5) -> float | None:
    """F0 par autocorrélation normalisée sur une trame ; None si non voisé.

    Retient le plus petit retard dont la corrélation atteint 90 % du maximum
    (limite les erreurs d'octave vers le bas).
    """
    start = frame_index * HOP_SIZE
    seg = list(samples[start:start + FRAME_SIZE])
    if len(seg) < FRAME_SIZE:
        return None
    mean = sum(seg) / FRAME_SIZE
    x = [s - mean for s in seg]
    lag_min = int(ANALYSIS_RATE / f0_max)
    lag_max = min(FRAME_SIZE - 2, int(ANALYSIS_RATE / f0_min))
    corr: list[tuple[int, float]] = []
    for lag in range(lag_min, lag_max + 1):
        num = 0.0
        e0 = 0.0
        e1 = 0.0
        for i in range(FRAME_SIZE - lag):
            a = x[i]
            b = x[i + lag]
            num += a * b
            e0 += a * a
            e1 += b * b
        den = math.sqrt(e0 * e1)
        corr.append((lag, num / den if den > EPS else 0.0))
    best = max(c for _, c in corr)
    if best < min_corr:
        return None
    for lag, c in corr:
        if c >= 0.9 * best:
            return ANALYSIS_RATE / lag
    return None


def lowpass_fir(cutoff_hz: float, rate: float, taps: int = 63) -> list[float]:
    """Filtre RIF passe-bas (sinus cardinal fenêtré Blackman), gain unitaire."""
    fc = cutoff_hz / rate
    mid = (taps - 1) / 2.0
    h = []
    for i in range(taps):
        m = i - mid
        sinc = 2.0 * fc if m == 0 else math.sin(2.0 * math.pi * fc * m) / (math.pi * m)
        w = 0.42 - 0.5 * math.cos(2.0 * math.pi * i / (taps - 1)) + 0.08 * math.cos(4.0 * math.pi * i / (taps - 1))
        h.append(sinc * w)
    total = sum(h)
    return [v / total for v in h]


def resample(samples: Sequence[float], rate_in: int, rate_out: int = ANALYSIS_RATE) -> list[float]:
    """Rééchantillonnage : passe-bas anti-repliement (si décimation) + interpolation linéaire.

    Suffisant pour des énergies de bande ; sur iOS, `AVAudioConverter` fait mieux
    et livre directement du 24 kHz mono au détecteur.
    """
    if rate_in == rate_out:
        return list(samples)
    src = list(samples)
    if rate_out < rate_in:
        h = lowpass_fir(0.45 * rate_out, rate_in)
        half = len(h) // 2
        filtered = []
        n = len(src)
        for i in range(n):
            acc = 0.0
            for k, hk in enumerate(h):
                j = i + half - k
                if 0 <= j < n:
                    acc += hk * src[j]
            filtered.append(acc)
        src = filtered
    ratio = rate_in / rate_out
    n_out = int(len(src) / ratio)
    out = []
    for i in range(n_out):
        pos = i * ratio
        j = int(pos)
        frac = pos - j
        a = src[j]
        b = src[j + 1] if j + 1 < len(src) else a
        out.append(a + (b - a) * frac)
    return out

File: eveil/ios/WordEndCore/Sources/WordEndCore/DSP.swift
// DSP.swift — primitives de traitement du signal, portage ligne à ligne de
// `eveil/reference/wordend/dsp.py` (la référence Python, testée).
//
// Règle de portabilité : mêmes constantes, mêmes ordres d'opérations, même
// PRNG. La parité est vérifiée par `GoldenVectorsTests` contre
// `golden_vectors.json`, produit par la référence Python.
//
// Swift pur (Foundation) : compile sur iOS, macOS et Linux.

import Foundation

/// Constantes d'analyse partagées (identiques à la référence Python).
public enum WordEnd {
    /// Fréquence d'analyse canonique : assez pour les sibilantes d'enfant
    /// (énergie jusque vers 10 kHz), deux fois moins coûteuse que 48 kHz.
    public static let analysisRate = 24_000
    /// Taille de trame = taille FFT (21,3 ms à 24 kHz).
    public static let frameSize = 512
    /// Pas d'analyse (10 ms à 24 kHz).
    public static let hopSize = 240
    /// Largeur d'un bin FFT : 46,875 Hz, exactement représentable.
    public static let binHz = Double(analysisRate) / Double(frameSize)
    /// Plancher numérique avant log10.
    public static let eps = 1e-12

    /// Pas d'analyse en millisecondes (10 ms).
    public static var frameMs: Double { 1000.0 * Double(hopSize) / Double(analysisRate) }

    /// Durée → nombre de trames. `round()` de Python arrondit au pair : on fait de même.
    public static func msToFrames(_ ms: Double) -> Int {
        Int((ms / frameMs).rounded(.toNearestOrEven))
    }
}

// MARK: - PRNG déterministe

/// SplitMix64 — identique bit à bit à la version Python (`&+`, `&*` = modulo 2⁶⁴).
public struct SplitMix64: Sendable {
    public private(set) var state: UInt64

    public init(seed: UInt64) { state = seed }

    public mutating func nextU64() -> UInt64 {
        state = state &+ 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }

    /// Uniforme dans [0, 1) sur 53 bits.
    public mutating func nextUnit() -> Double {
        Double(nextU64() >> 11) * (1.0 / 9_007_199_254_740_992.0)
    }

    /// Uniforme dans [-1, 1) — bruit blanc sans fonction transcendante.
    public mutating func nextSigned() -> Double { 2.0 * nextUnit() - 1.0 }
}

// MARK: - FFT portable

/// FFT radix-2 itérative (Cooley-Tukey) d'un signal réel, en Double.
///
/// À 512 points × 100 trames/s, le coût est négligeable sur n'importe quel iPad :
/// Accelerate n'est pas nécessaire au budget CPU. Le backend vDSP reste
/// disponible (`AccelerateSpectrum`) et vérifié par un test de parité.
public struct RadixTwoFFT: Sendable {
    public let size: Int
    private let cosTable: [Double]
    private let sinTable: [Double]

    public init(size: Int) {
        precondition(size > 0 && size & (size - 1) == 0, "taille FFT non puissance de 2 : \(size)")
        self.size = size
        var c: [Double] = []
        var s: [Double] = []
        c.reserveCapacity(size / 2)
        s.reserveCapacity(size / 2)
        for k in 0..<(size / 2) {
            let ang = -2.0 * Double.pi * Double(k) / Double(size)
            c.append(cos(ang))
            s.append(sin(ang))
        }
        cosTable = c
        sinTable = s
    }

    /// DFT de `input` (réel, `size` échantillons) → parties réelle et imaginaire.
    public func transform(_ input: [Double]) -> (re: [Double], im: [Double]) {
        let n = size
        precondition(input.count == n, "entrée de \(input.count) échantillons pour une FFT de \(n)")
        var re = input
        var im = [Double](repeating: 0.0, count: n)
        var j = 0
        for i in 1..<max(n, 1) {                    // permutation bit-reverse
            var bit = n >> 1
            while j & bit != 0 {
                j ^= bit
                bit >>= 1
            }
            j ^= bit
            if i < j {
                re.swapAt(i, j)
                im.swapAt(i, j)
            }
        }
        var len = 2
        while len <= n {
            let half = len / 2
            let step = n / len
            var start = 0
            while start < n {
                for k in 0..<half {
                    let wr = cosTable[k * step]
                    let wi = sinTable[k * step]
                    let xr = re[start + k + half]
                    let xi = im[start + k + half]
                    let vr = xr * wr - xi * wi
                    let vi = xr * wi + xi * wr
                    let ur = re[start + k]
                    let ui = im[start + k]
                    re[start + k] = ur + vr
                    im[start + k] = ui + vi
                    re[start + k + half] = ur - vr
                    im[start + k + half] = ui - vi
                }
                start += len
            }
            len *= 2
        }
        return (re, im)
    }

    /// Spectre de puissance |X[k]|² pour k = 0 … size/2.
    public func powerSpectrum(_ input: [Double]) -> [Double] {
        let (re, im) = transform(input)
        return (0...(size / 2)).map { re[$0] * re[$0] + im[$0] * im[$0] }
    }
}

// MARK: - Descripteurs par trame

/// Bande de fréquences [lo, hi) en Hz.
public struct Band: Equatable, Sendable {
    public var lo: Double
    public var hi: Double
    public init(_ lo: Double, _ hi: Double) {
        self.lo = lo
        self.hi = hi
    }
}

/// Bandes d'analyse. Valeurs par défaut justifiées dans docs/EVEIL.md §5.
public struct Bands: Equatable, Sendable {
    public var low = Band(300.0, 2500.0)        // voyelles (F1/F2)
    public var high = Band(3000.0, 11000.0)     // bruit de friction sibilant
    public var centroid = Band(1000.0, 11000.0)
    public init() {}
}

/// Indices de bins [k0, k1) couvrant la bande.
public func bandBins(_ band: Band) -> (Int, Int) {
    let k0 = max(1, Int(ceil(band.lo / WordEnd.binHz)))
    let k1 = min(WordEnd.frameSize / 2 + 1, Int(ceil(band.hi / WordEnd.binHz)))
    return (k0, max(k0, k1))
}

@inline(__always) func toDb(_ power: Double) -> Double { 10.0 * log10(power + WordEnd.eps) }

public struct FrameFeatures: Equatable, Sendable {
    public let index: Int
    public let rmsDb: Double       // énergie totale (dB, référence arbitraire)
    public let lowDb: Double       // énergie bande « voyelle »
    public let highDb: Double      // énergie bande « friction »
    public let highRatio: Double   // E_high / (E_low + E_high) ∈ [0, 1]
    public let centroidHz: Double  // centroïde spectral sur la bande `centroid`
    public let zcrHz: Double       // passages par zéro par seconde
    public let peak: Double        // max |x| (écrêtage)
}

/// Fenêtre de Hann périodique (dénominateur N), comme la référence.
public let hannWindow: [Double] = (0..<WordEnd.frameSize).map {
    0.5 - 0.5 * cos(2.0 * Double.pi * Double($0) / Double(WordEnd.frameSize))
}

/// Nombre de trames entièrement contenues dans le signal.
public func frameCount(_ sampleCount: Int) -> Int {
    sampleCount < WordEnd.frameSize ? 0 : 1 + (sampleCount - WordEnd.frameSize) / WordEnd.hopSize
}

/// Calcule les descripteurs de trames. Réutilisable (FFT pré-calculée).
public struct FeatureExtractor: Sendable {
    public let bands: Bands
    private let fft = RadixTwoFFT(size: WordEnd.frameSize)
    private let low: (Int, Int)
    private let high: (Int, Int)
    private let centroid: (Int, Int)

    public init(bands: Bands = Bands()) {
        self.bands = bands
        low = bandBins(bands.low)
        high = bandBins(bands.high)
        centroid = bandBins(bands.centroid)
    }

    public func frame(_ samples: [Double], index: Int) -> FrameFeatures {
        let n = WordEnd.frameSize
        let start = index * WordEnd.hopSize
        var seg = [Double](repeating: 0.0, count: n)
        let end = min(samples.count, start + n)
        if start < end {
            for i in start..<end { seg[i - start] = samples[i] }
        }
        var total = 0.0
        for v in seg { total += v }
        let mean = total / Double(n)
        let x = seg.map { $0 - mean }
        var energy = 0.0
        for v in x { energy += v * v }
        energy /= Double(n)

        var crossings = 0
        var prevNeg = x[0] < 0.0
        for i in 1..<n {
            let neg = x[i] < 0.0
            if neg != prevNeg { crossings += 1 }
            prevNeg = neg
        }
        let zcr = Double(crossings) * Double(WordEnd.analysisRate) / Double(n)
        var peak = 0.0
        for v in seg { peak = max(peak, abs(v)) }

        var windowed = [Double](repeating: 0.0, count: n)
        for i in 0..<n { windowed[i] = x[i] * hannWindow[i] }
        let power = fft.powerSpectrum(windowed)

        var eLow = 0.0
        for k in low.0..<low.1 { eLow += power[k] }
        var eHigh = 0.0
        for k in high.0..<high.1 { eHigh += power[k] }
        var eC = 0.0
        var eCF = 0.0
        for k in centroid.0..<centroid.1 {
            eC += power[k]
            eCF += power[k] * Double(k) * WordEnd.binHz
        }
        let centroidHz = eC > WordEnd.eps ? eCF / eC : 0.0
        let ratio = (eLow + eHigh) > WordEnd.eps ? eHigh / (eLow + eHigh) : 0.0
        return FrameFeatures(index: index, rmsDb: toDb(energy), lowDb: toDb(eLow), highDb: toDb(eHigh),
                             highRatio: ratio, centroidHz: centroidHz, zcrHz: zcr, peak: peak)
    }

    public func frames(_ samples: [Double]) -> [FrameFeatures] {
        (0..<frameCount(samples.count)).map { frame(samples, index: $0) }
    }
}

// MARK: - Outils statistiques et signaux

/// Percentile « rang inférieur » déterministe (p ∈ [0, 1]).
public func percentile(_ values: [Double], _ p: Double) -> Double {
    precondition(!values.isEmpty, "percentile d'une liste vide")
    let ordered = values.sorted()
    let idx = Int(p * Double(ordered.count - 1))
    return ordered[max(0, min(ordered.count - 1, idx))]
}

/// Moyenne glissante centrée ; aux bords, moyenne des valeurs disponibles.
public func movingAverage(_ values: [Double], width: Int) -> [Double] {
    let half = width / 2
    let n = values.count
    var out: [Double] = []
    out.reserveCapacity(n)
    for t in 0..<n {
        let lo = max(0, t - half)
        let hi = min(n, t + half + 1)
        var s = 0.0
        for i in lo..<hi { s += values[i] }
        out.append(s / Double(hi - lo))
    }
    return out
}

/// F0 par autocorrélation normalisée sur une trame ; nil si non voisé.
/// Retient le plus petit retard dont la corrélation atteint 90 % du maximum.
public func estimateF0(_ samples: [Double], frameIndex: Int, f0Min: Double = 90.0,
                       f0Max: Double = 800.0, minCorr: Double = 0.5) -> Double? {
    let n = WordEnd.frameSize
    let start = frameIndex * WordEnd.hopSize
    guard start >= 0, start + n <= samples.count else { return nil }
    var total = 0.0
    for i in start..<(start + n) { total += samples[i] }
    let mean = total / Double(n)
    let x = (start..<(start + n)).map { samples[$0] - mean }
    let lagMin = Int(Double(WordEnd.analysisRate) / f0Max)
    let lagMax = min(n - 2, Int(Double(WordEnd.analysisRate) / f0Min))
    guard lagMin <= lagMax else { return nil }
    var corr: [(Int, Double)] = []
    for lag in lagMin...lagMax {
        var num = 0.0
        var e0 = 0.0
        var e1 = 0.0
        for i in 0..<(n - lag) {
            let a = x[i]
            let b = x[i + lag]
            num += a * b
            e0 += a * a
            e1 += b * b
        }
        let den = (e0 * e1).squareRoot()
        corr.append((lag, den > WordEnd.eps ? num / den : 0.0))
    }
    let best = corr.map { $0.1 }.max() ?? 0.0
    guard best >= minCorr else { return nil }
    for (lag, c) in corr where c >= 0.9 * best {
        return Double(WordEnd.analysisRate) / Double(lag)
    }
    return nil
}

/// Filtre RIF passe-bas (sinus cardinal fenêtré Blackman), gain unitaire.
public func lowpassFIR(cutoffHz: Double, rate: Double, taps: Int = 63) -> [Double] {
    let fc = cutoffHz / rate
    let mid = Double(taps - 1) / 2.0
    var h: [Double] = []
    for i in 0..<taps {
        let m = Double(i) - mid
        let sinc = m == 0 ? 2.0 * fc : sin(2.0 * Double.pi * fc * m) / (Double.pi * m)
        let w = 0.42 - 0.5 * cos(2.0 * Double.pi * Double(i) / Double(taps - 1))
            + 0.08 * cos(4.0 * Double.pi * Double(i) / Double(taps - 1))
        h.append(sinc * w)
    }
    var total = 0.0
    for v in h { total += v }
    return h.map { $0 / total }
}

/// Rééchantillonnage : passe-bas anti-repliement (si décimation) + interpolation
/// linéaire. Sur iOS, `AVAudioConverter` livre déjà du 24 kHz mono : cette
/// fonction sert aux tests, au banc et aux plateformes sans AVFoundation.
public func resample(_ samples: [Double], from rateIn: Int, to rateOut: Int = WordEnd.analysisRate) -> [Double] {
    if rateIn == rateOut { return samples }
    var src = samples
    if rateOut < rateIn {
        let h = lowpassFIR(cutoffHz: 0.45 * Double(rateOut), rate: Double(rateIn))
        let half = h.count / 2
        let n = src.count
        var filtered = [Double](repeating: 0.0, count: n)
        for i in 0..<n {
            var acc = 0.0
            for (k, hk) in h.enumerated() {
                let j = i + half - k
                if j >= 0 && j < n { acc += hk * src[j] }
            }
            filtered[i] = acc
        }
        src = filtered
    }
    let ratio = Double(rateIn) / Double(rateOut)
    let nOut = Int(Double(src.count) / ratio)
    var out: [Double] = []
    out.reserveCapacity(nOut)
    for i in 0..<nOut {
        let pos = Double(i) * ratio
        let j = Int(pos)
        let frac = pos - Double(j)
        let a = src[j]
        let b = j + 1 < src.count ? src[j + 1] : a
        out.append(a + (b - a) * frac)
    }
    return out
}
